sql/wsrep_trans_observer.h

```diff
@@ -373,6 +373,7 @@ static inline int wsrep_after_commit(THD *thd, bool all) {
     thread handler to register in wsrep group commit queue but since storage
     engine commit is not done it would fail to unregister the said thread
     handler as part of storage engine commit. Handle unregistration here. */
+    wsrep_wait_for_turn_in_group_commit(thd);
```

Comment:
Now that we are touching this part of the code, I think it would better to fix the other issues found in this code as well.

1. In https://github.com/percona/percona-xtradb-cluster/blob/8.0/sql/wsrep_binlog.cc#L474-L475, 
  the order of the call to `mysql_mutex_unlock(&LOCK_wsrep_group_commit)` and `mysql_cond_broadcast(&COND_wsrep_group_commit)` seem to be wrong. The correct order is 
```
  mysql_cond_broadcast(&COND_wsrep_group_commit);
  mysql_mutex_unlock(&LOCK_wsrep_group_commit);
```


2. Can you please add an assert that the current thread is in the front of the queue when it is trying to unregister from the queue.
https://github.com/percona/percona-xtradb-cluster/blob/8.0/sql/wsrep_binlog.cc#L469

something like 
```
  thd->wsrep_enforce_group_commit = false;
  assert(wsrep_group_commit_queue.front() == thd);
  wsrep_group_commit_queue.pop();
```

3. Can we use mutex lock guards in all occurences where we acquire `LOCK_wsrep_group_commit`

```
#include "mutex_lock.h" and use
MUTEX_LOCK (lock_guard, &LOCK_wsrep_group_commit)
instead of
mysql_mutex_lock(&LOCK_wsrep_group_commit);
```

4. This is optional, but it would be better to have `DBUG_TRACE` and `return` statements instead of combinations of `DBUG_ENTER` and `DBUG_VOID_RETURN`

Reply:
Hi @venkatesh-prasad-v ,

thanks for your review. I had look on your suggestions and did a cleanup like you suggested.
Please have a look again.

Best,
Kevin